sources/api/apiclient/src/exec.rs

```diff
@@ -0,0 +1,683 @@
+    // separate threads, which simplifies the use of blocking calls, not requiring a totally new
+    // async infrastructure.
+    let (write, read) = ws_stream.split();
+
+    // We make a multi-producer channel that forwards anything it receives to the WebSocket; we can
+    // share the transmission end of the channel with any number of threads that need to send
+    // messages to the server.
+    let (ws_tx, ws_rx) = mpsc::unbounded();
+    let forward_to_ws = ws_rx.map(Ok).forward(write);
+    debug!("Spawning task to write to WebSocket");
+    tokio::spawn(forward_to_ws);
+
+    // The first thing we want to send is an initialize message that tells the server what program
+    // we want to run, what container to run it in, and whether we want a TTY.  (It's important
+    // not to send other types of messages first or the server won't have a process to act on and
+    // will reject us.  It'd be nice to send initialization parameters in the HTTP request body,
+    // but not all WebSocket clients support it.)
+    debug!(
+        "Sending initialize request for target '{}' with tty: {} and command: {:?}",
+        target,
+        terminal.tty().is_some(),
+        command
+    );
+    let init = Initialize {
+        command,
+        target,
+        tty: terminal.tty().clone(),
+    };
+    // Control messages go to the server in a text channel, so we serialize to JSON before sending.
+    let msg = serde_json::to_string(&ClientMessage::Initialize(init)).context(error::Serialize)?;
+    ws_tx
+        .unbounded_send(Message::Text(msg))
+        .context(error::SendMessage {
+            kind: "initialization",
+        })?;
+
+    // Now that the server knows what we want, we set up helper threads to manage communication.
+    // First, a heartbeat type that regularly pings the server and keeps track of responses.
+    let mut heartbeat = Heartbeat::new(ws_tx.clone());
+    // Next, a type that watches for signals to the local process, and either forwards them to the
+    // server (e.g. if you change your window size) or ends communication (e.g. for SIGTERM).
+    let mut signal_handler = HandleSignals::new(ws_tx.clone())?;
+
+    // We don't want to overload the server with our process input.  It sends us capacity updates
+    // to let us know how many more messages we can send before we should wait.  We keep track of
+    // that capacity in an AtomicCapacity that we can share across threads - the one where we
+    // receive messages from the server, and the one that's reading input to send to the server.
+    let capacity = Arc::new(AtomicCapacity::default());
+    let capacity_reader = Arc::clone(&capacity);
+
+    // Start a thread that reads input from the user and sends it across the WebSocket, waiting for
+    // capacity between reads if necessary.
+    let _ = ReadFromUser::new(ws_tx.clone(), capacity_reader, terminal.tty().is_some());
```

Comment:
I'm not sure I follow what you think is reckless...  We don't have to infer that ReadFromUser is done because it doesn't really matter if it's done except in the case that we read an EOF, in which case we send the server a ContentComplete; in any other case, we want to funnel input to the server forever..?

Reply:
In combination with https://github.com/bottlerocket-os/bottlerocket/pull/1802#discussion_r742310304 I think I see what you're saying; we can end early in the case where we receive an "impossible" count from the server or an "impossible" read in the TTY case.  I'll try it.